src/query/expression/src/types/map.rs

```diff
@@ -298,110 +297,110 @@ unsafe impl<'a, K: ValueType, V: ValueType> TrustedLen for KvIterator<'a, K, V>
 // Structuarally equals to `Array(Tuple(K, V))` but treated distinct from `Array(Tuple(K, V))`
 // in unification.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct MapType<T: ValueType>(PhantomData<T>);
+pub struct MapType<K: ValueType, V: ValueType>(PhantomData<(K, V)>);
 
-pub type MapInternal<T> = ArrayType<KvPair<StringType, T>>;
+pub type MapInternal<K, V> = ArrayType<KvPair<K, V>>;
 
-impl<T: ValueType> ValueType for MapType<T> {
-    type Scalar = <MapInternal<T> as ValueType>::Scalar;
-    type ScalarRef<'a> = <MapInternal<T> as ValueType>::ScalarRef<'a>;
-    type Column = <MapInternal<T> as ValueType>::Column;
-    type Domain = <MapInternal<T> as ValueType>::Domain;
-    type ColumnIterator<'a> = <MapInternal<T> as ValueType>::ColumnIterator<'a>;
-    type ColumnBuilder = <MapInternal<T> as ValueType>::ColumnBuilder;
+impl<K: ValueType, V: ValueType> ValueType for MapType<K, V> {
+    type Scalar = <MapInternal<K, V> as ValueType>::Scalar;
+    type ScalarRef<'a> = <MapInternal<K, V> as ValueType>::ScalarRef<'a>;
+    type Column = <MapInternal<K, V> as ValueType>::Column;
+    type Domain = <MapInternal<K, V> as ValueType>::Domain;
+    type ColumnIterator<'a> = <MapInternal<K, V> as ValueType>::ColumnIterator<'a>;
+    type ColumnBuilder = <MapInternal<K, V> as ValueType>::ColumnBuilder;
 
     #[inline]
     fn upcast_gat<'short, 'long: 'short>(long: Self::ScalarRef<'long>) -> Self::ScalarRef<'short> {
-        <MapInternal<T> as ValueType>::upcast_gat(long)
+        <MapInternal<K, V> as ValueType>::upcast_gat(long)
     }
 
     fn to_owned_scalar<'a>(scalar: Self::ScalarRef<'a>) -> Self::Scalar {
-        <MapInternal<T> as ValueType>::to_owned_scalar(scalar)
+        <MapInternal<K, V> as ValueType>::to_owned_scalar(scalar)
     }
 
     fn to_scalar_ref<'a>(scalar: &'a Self::Scalar) -> Self::ScalarRef<'a> {
-        <MapInternal<T> as ValueType>::to_scalar_ref(scalar)
+        <MapInternal<K, V> as ValueType>::to_scalar_ref(scalar)
     }
 
     fn try_downcast_scalar<'a>(scalar: &'a ScalarRef) -> Option<Self::ScalarRef<'a>> {
-        <MapInternal<T> as ValueType>::try_downcast_scalar(scalar)
+        <MapInternal<K, V> as ValueType>::try_downcast_scalar(scalar)
     }
 
     fn try_downcast_column<'a>(col: &'a Column) -> Option<Self::Column> {
-        <MapInternal<T> as ValueType>::try_downcast_column(col)
+        <MapInternal<K, V> as ValueType>::try_downcast_column(col)
```

Comment:
Should downcast `Column::Map`, but not `Column::Array`

Reply:
`Column::Map` is also an `ArrayColumn`.

Reply:
MapInternal will downcast `Column::Array` instead of `Column::Map`

Reply:
ok